# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/fs/shared/ManagedMozPromiseRequestHolder.h
# Commit: cb35e5c2851c
# Full Hash: cb35e5c2851c84fb955b7c94ae209824a20b7d15
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2023-01-17 16:13:02
# Description:
#   Bug 1806490 - Make it possible to track all CreateFileSystemManagerChild requests in FileSystemManager; r=dom-storage-reviewers,jari
#   
#   It can happen that BeginRequest is called multiple times when the actor doesn't
#   exist yet. Such situations can't be handled by a single MozPromiseRequestHolder.
#   
# ==============================================================================

diff -r e939a80b5b75 -r cb35e5c2851c dom/fs/shared/ManagedMozPromiseRequestHolder.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/dom/fs/shared/ManagedMozPromiseRequestHolder.h	Mon Jan 16 14:46:20 2023 +0000
@@ -0,0 +1,37 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef DOM_FS_SHARED_MANAGEDMOZPROMISEREQUESTHOLDER_H_
+#define DOM_FS_SHARED_MANAGEDMOZPROMISEREQUESTHOLDER_H_
+
+#include "mozilla/MozPromise.h"
+#include "mozilla/RefPtr.h"
+#include "nsISupportsImpl.h"
+
+namespace mozilla::dom::fs {
+
+template <typename Manager, typename PromiseType>
+class ManagedMozPromiseRequestHolder final
+    : public MozPromiseRequestHolder<PromiseType> {
+ public:
+  explicit ManagedMozPromiseRequestHolder(Manager* aManager)
+      : mManager(aManager) {
+    mManager->RegisterPromiseRequestHolder(this);
+  }
+
+  NS_INLINE_DECL_REFCOUNTING(ManagedMozPromiseRequestHolder)
+
+ private:
+  ~ManagedMozPromiseRequestHolder() {
+    mManager->UnregisterPromiseRequestHolder(this);
+  }
+
+  RefPtr<Manager> mManager;
+};
+
+}  // namespace mozilla::dom::fs
+
+#endif  // DOM_FS_SHARED_MANAGEDMOZPROMISEREQUESTHOLDER_H_